osgp/platform/osgp-adapter-domain-core/src/main/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementService.java

```diff
@@ -371,38 +368,159 @@ private void checkFirmwareHistory(final String deviceId,
         return firmwareVersionsToCheck;
     }
 
-    public void tryToAddFirmwareVersionToHistory(final String deviceIdentification,
-            final FirmwareVersion firmwareVersion) {
+    /**
+     * @param deviceId
+     *            the id of the device we are checking
+     * @param firmwareVersions
+     *            the list of firmware modules versions (so type and version) to
+     *            check if they are currently installed on the device, using the
+     *            history of the devices firmware history
+     * @return a list of firmware module versions not present in the the devices
+     *         firmware history
+     * @throws FunctionalException
+     */
+    public List<FirmwareVersion> checkFirmwareHistoryForModuleVersionsNotCurrentlyInstalled(final String deviceId,
+            final List<FirmwareVersion> firmwareVersions) {
+
+        if (firmwareVersions.isEmpty()) {
+            return firmwareVersions;
+        }
+        // copy input parameter
+        final List<FirmwareVersion> firmwareVersionsToCheck = new ArrayList<>();
+        firmwareVersionsToCheck.addAll(firmwareVersions);
+
+        // gets list of all historically installed modules
+        final Device device = this.deviceRepository.findByDeviceIdentification(deviceId);
+        final List<DeviceFirmwareFile> deviceFirmwareFiles = this.deviceFirmwareFileRepository
+                .findByDeviceOrderByInstallationDateAsc(device);
+
+        // Transform this list so it contains only the latest entry for each
+        // moduleType
+        Map<String, FirmwareVersionWithInstallationDate> currentlyInstalledFirmwareVersionsPerType = new HashMap<String, FirmwareVersionWithInstallationDate>();
+
+        for (DeviceFirmwareFile firmwareFile : deviceFirmwareFiles) {
+
+            Map<FirmwareModule, String> fwms = firmwareFile.getFirmwareFile().getModuleVersions();
+            Date installationDate = firmwareFile.getInstallationDate();
+
+            for (Map.Entry<FirmwareModule, String> entry : fwms.entrySet()) {
+                String version = entry.getValue();
+                FirmwareModule fwm = entry.getKey();
+                // check if this installation of this same kind of module is
+                // of a later date
+                if (currentlyInstalledFirmwareVersionsPerType.containsKey(fwm.getDescription())
+                        && currentlyInstalledFirmwareVersionsPerType.get(fwm.getDescription())
+                                .getInstallationDate()
+                                .before(installationDate)) {
+                    currentlyInstalledFirmwareVersionsPerType.replace(fwm.getDescription(),
+                            new FirmwareVersionWithInstallationDate(installationDate, new FirmwareVersion(
+                                    FirmwareModuleType.forDescription(fwm.getDescription()), version)));
```

Comment:
I think replacing the entry will fix the problem reported concerning a previously installed version. For example when switching back to an older firmware version.